src/de.hpi.swa.graal.squeak/src/de/hpi/swa/graal/squeak/nodes/process/RemoveProcessFromListNode.java

```diff
@@ -32,10 +32,12 @@ public final void executeRemove(final PointersObject process, final PointersObje
 
     @Specialization(guards = "process == first")
     protected final void doRemoveEqual(final PointersObject process, final PointersObject list, @SuppressWarnings("unused") final PointersObject first, final AbstractSqueakObject last) {
-        final Object next = readNode.execute(process, PROCESS.NEXT_LINK);
-        writeNode.execute(list, LINKED_LIST.FIRST_LINK, next);
         if (process == last) {
             writeNode.executeNil(list, LINKED_LIST.LAST_LINK);
+            writeNode.executeNil(list, LINKED_LIST.FIRST_LINK);
+        } else {
+            final Object next = readNode.execute(process, PROCESS.NEXT_LINK);
+            writeNode.execute(list, LINKED_LIST.FIRST_LINK, next);
```

Comment:
When is this an issue? And why isn't it a problem in the reference VM? This is the corresponding Slang code:
```smalltalk
[nextLink := objectMemory fetchPointer: NextLinkIndex ofObject: aProcess.
 self deny: (objectMemory isForwarded: nextLink).
 objectMemory storePointer: FirstLinkIndex ofObject: aList withValue: nextLink.
 aProcess  = lastLink ifTrue:
	[objectMemory storePointerUnchecked: LastLinkIndex ofObject: aList withValue: objectMemory nilObject]]
```

Reply:
It is a problem because by leaving the first link not nil we are invalidating the check for an empty list (which relies on the first link being nil), which in turn is used in WakeHighestPriorityNode.
As a result, we end up trying to wake up the wrong process (from what should have been an empty list)

Reply:
But why is this not a problem in the corresponding Slang code and OpenSmalltalkVM?

Reply:
I have seen it happen. And isEmptyList is also used in SignalSemaphoreNode.
I don't know what to say about Slang, but it clearly does not respect the LinkedList protocol - it could be a bug

Reply:
I think this should trigger the bug in Squeak:
| p | p := [Transcript show: 'bla1'] newProcess priority: 37. p resume. p suspend
But it doesn't. I assume the checks happening within the call to #isLiveContext: are sufficient to filter it out. Note that we do not do those extensive checks, we only guard against a nil suspendedContext